### PR TITLE
Add a workaround for inner rects on transformed clips.

### LINKED ISSUE #1814: Transformed images with non-transformed clips (and no transformed clip) are not clipped
https://bugzilla.mozilla.org/show_bug.cgi?id=1405951

Expected vs actual:
screen shot 2017-10-05 at 2 33 02 pm

In the following YAML testcase, I have a transformed stacking context, and two clips: one clip outside the stacking context (`id: 1`) and one clip inside the stacking context (`id: 2`). The transformed clip `2` has the untransformed clip `1` as its parent clip.

I also have an image inside the transformed stacking context. When I set the image's `clip-and-scroll` to `1`, it is drawn unclipped.
If I set the image's `clip-and-scroll` to `2`, then the image is properly clipped by both `2` and `1`.

```yaml
---
root: 
  items:
    -
      bounds: [16, 16, 404, 404]
      "clip-rect": [16, 16, 404, 404]
      "clip-and-scroll": 0
      "backface-visible": true
      type: border
      width: 2
      "border-type": normal
      color: black
      style: solid
    -
      bounds: [18, 18, 400, 200]
      "clip-rect": [18, 18, 400, 400]
      "clip-and-scroll": 0
      "backface-visible": true
      type: clip
      id: 1
      "content-size": [400, 400]
    -
      bounds: [0, 0, 0, 0]
      "clip-rect": [0, 0, 0, 0]
      "clip-and-scroll": 1
      "backface-visible": true
      type: "stacking-context"
      "scroll-policy": scrollable
      transform: [0.70710677, 0.70710677, 0, 0, -0.70710677, 0.70710677, 0, 0, 0, 0, 1, 0, 218, -64.84271, 0, 1]
      "transform-style": flat
      items:
        -
          bounds: [18, 18, 400, 400]
          "clip-rect": [18, 18, 400, 400]
          "clip-and-scroll": 1
          "backface-visible": true
          type: clip
          id: 2
          "content-size": [400, 400]
        - 
          bounds: [0, 0, 400, 400]
          "clip-rect": [0, 0, 400, 400]
          "clip-and-scroll": 1
          "backface-visible": true
          image: "res/1507226429-img-91.png"
          "stretch-size": [400, 400]
          "tile-spacing": [400, 400]
    - 
      bounds: [0, 0, 0, 0]
      "clip-rect": [0, 0, 0, 0]
      "clip-and-scroll": 0
      "backface-visible": true
  id: [1, 1]
pipelines: []
```

`res/1507226429-img-91.png`: 
1507226429-img-91

### PR DESCRIPTION
**Pull request by glennw**
Fixes #1814.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1820)

### CODE REVIEW

**Comment by glennw** [comment]
(Still needs a try run before merge)

r? @kvark

**Comment by glennw** [comment]
Servo / wrench tests pass.

Gecko try looks ok, I think (two unexpected passes from the recent transform change, and two jobs that didn't start correct - re-running those now).

https://hg.mozilla.org/try/rev/0a457dc9cac0a58e0cf922b8a61e5849b4b67066
https://treeherder.mozilla.org/#/jobs?repo=try&revision=0a457dc9cac0a58e0cf922b8a61e5849b4b67066

File: webrender/src/render_task.rs [hunk 1]
```diff
@@ -329,7 +329,10 @@ impl RenderTask {
                 }
 
                 match clip_info.bounds.inner {
-                    Some(ref inner) if !inner.device_rect.is_empty() => {
+                    // Inner rects aren't valid if the item is not axis-aligned, which can
+                    // be determined by the apply_rectangles field. This is mostly a band-aid
+                    // until we have better handling of inner rectangles for transformed clips.
+                    Some(ref inner) if !work_item.apply_rectangles && !inner.device_rect.is_empty() => {
```
Comment (kvark) [inline webrender/src/render_task.rs]: weird, I believe they should be valid

**Comment by kvark** [comment]
I'm worried that band-aiding and working around the malfunctioning parts of WR in order to suffice the reftests would make us forget to actually make it work properly. We need to either create new bugs to keep track of it, leave `TODO` comments (as a bare minimum), or not close WR-side bugs until proper solutions are in place (as opposed to workarounds), while still closing the corresponding Gecko issues.

**Pull request merged**